irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -46,85 +47,117 @@ namespace iroha {
+            return std::unique_ptr<ametsuchi::MutableStorage>{};
+          });
+    }
 
-      if (validator_->validateBlock(*commit_message, *storage)) {
-        // Block can be applied to current storage
-        // Commit to main Ametsuchi
-        mutableFactory_->commit(std::move(storage));
+    void SynchronizerImpl::processApplicableBlock(
+        const shared_model::interface::BlockVariant &committed_block_variant)
+        const {
+      iroha::visit_in_place(
+          committed_block_variant,
+          [&](std::shared_ptr<shared_model::interface::Block> block_ptr) {
+            auto storage = createTemporaryStorage();
+            if (not storage) {
+              return;
+            }
+            storage->apply(*block_ptr, trueStorageApplyPredicate);
+            mutable_factory_->commit(std::move(storage));
 
-        auto single_commit = rxcpp::observable<>::just(commit_message);
+            notifier_.get_subscriber().on_next(
+                rxcpp::observable<>::just(block_ptr));
+          },
+          [this](std::shared_ptr<shared_model::interface::EmptyBlock>
+                     empty_block_ptr) {
+            notifier_.get_subscriber().on_next(
+                rxcpp::observable<>::empty<
+                    std::shared_ptr<shared_model::interface::Block>>());
+          });
+    }
 
-        notifier_.get_subscriber().on_next(single_commit);
-      } else {
-        // Block can't be applied to current storage
-        // Download all missing blocks
-        for (const auto &signature : commit_message->signatures()) {
-          auto storageResult = mutableFactory_->createMutableStorage();
-          std::unique_ptr<ametsuchi::MutableStorage> storage;
-          storageResult.match(
-              [&](expected::Value<std::unique_ptr<ametsuchi::MutableStorage>>
-                      &_storage) { storage = std::move(_storage.value); },
-              [&](expected::Error<std::string> &error) {
-                storage = nullptr;
-                log_->error(error.error);
-              });
-          if (not storage) {
-            return;
-          }
-          auto network_chain = blockLoader_->retrieveBlocks(
-              shared_model::crypto::PublicKey(signature.publicKey()));
-          // Check chain last commit
+    void SynchronizerImpl::downloadAndApplyMissingChain(
+        const shared_model::interface::BlockVariant &committed_block_variant,
+        std::unique_ptr<ametsuchi::MutableStorage> storage) const {
+      while (true) {
+        for (const auto &signature : committed_block_variant.signatures()) {
           std::vector<std::shared_ptr<shared_model::interface::Block>> blocks;
-          network_chain.as_blocking().subscribe(
-              [&blocks](auto block) { blocks.push_back(block); });
-          auto is_chain_end_expected =
-              blocks.back()->hash() == commit_message->hash();
+          block_loader_
+              ->retrieveBlocks(
```

Comment:
Please consider not downloading whole chain

Reply:
It will not. `retrieveBlocks(..)` automatically starts to download blocks from current top